src/EventStore.Core/Services/Replication/ReplicationTrackingService.cs

```diff
@@ -0,0 +1,167 @@
+
+namespace EventStore.Core.Services.Replication {
+
+	public class ReplicationTrackingService :
+		IHandle<SystemMessage.StateChangeMessage>,
+		IHandle<SystemMessage.BecomeShuttingDown>,
+		IHandle<SystemMessage.SystemInit>,
+		IHandle<ReplicationTrackingMessage.ReplicaWriteAck>,
+		IHandle<ReplicationTrackingMessage.WriterCheckpointFlushed>,
+		IHandle<ReplicationTrackingMessage.MasterReplicatedTo> {
+		private readonly ILogger _log = LogManager.GetLoggerFor<ReplicationTrackingService>();
+		private readonly IPublisher _publisher;
+		private readonly ICheckpoint _replicationCheckpoint;
+		private readonly ICheckpoint _writerCheckpoint;
+		private readonly int _quorumSize;
+		private Thread _thread;
+		private bool _stop;
+		private VNodeState _state;
+		private long _idle = 1;
+		private readonly ConcurrentDictionary<Guid, long> _replicaLogPositions = new ConcurrentDictionary<Guid, long>();
+
+		private readonly ManualResetEventSlim _replicationChange = new ManualResetEventSlim(false, 1);
+		private readonly TaskCompletionSource<object> _tcs = new TaskCompletionSource<object>();
+
+		public Task Task {
+			get { return _tcs.Task; }
+		}
+
+		public ReplicationTrackingService(
+			IPublisher publisher,
+			int clusterNodeCount,
+			ICheckpoint replicationCheckpoint,
+			ICheckpoint writerCheckpoint) {
+			Ensure.NotNull(publisher, nameof(publisher));
+			Ensure.NotNull(replicationCheckpoint, nameof(replicationCheckpoint));
+			Ensure.NotNull(writerCheckpoint, nameof(writerCheckpoint));
+			Ensure.Positive(clusterNodeCount, nameof(clusterNodeCount));
+			_publisher = publisher;
+			_replicationCheckpoint = replicationCheckpoint;
+			_writerCheckpoint = writerCheckpoint;
+			_quorumSize = clusterNodeCount / 2 + 1;
+		}
+
+		public void Start() {
+			_thread = new Thread(TrackReplication) { IsBackground = true, Name = nameof(ReplicationTrackingService) };
+			_thread.Start();
+		}
+
+		public void Stop() {
+			_stop = true;
+		}
+		public bool IsIdle() { return Interlocked.Read(ref _idle) == 1; }
+		private void TrackReplication() {
+			long publishedPosition = 0;
+			try {
+				while (!_stop) {
+					//todo: consider if this will be too spammy, should we have a throttle here?
+					_replicationChange.Reset();
+					if (_state == VNodeState.Master) {
+						//Publish Log Commit Position
+						var newPos = _replicationCheckpoint.Read();
+						if (newPos > publishedPosition) {
+							_publisher.Publish(new ReplicationTrackingMessage.ReplicatedTo(newPos));
+							publishedPosition = newPos;
+						}
+					}
+					_idle = 1;
+					_replicationChange.Wait(100);
+					_idle = 0;
+				}
+			} catch (Exception exc) {
+				_log.FatalException(exc, $"Error in {nameof(ReplicationTrackingService)}. Terminating...");
+				_tcs.TrySetException(exc);
+				Application.Exit(ExitCode.Error,
+					$"Error in {nameof(ReplicationTrackingService)}. Terminating...\nError: " + exc.Message);
+				//todo: is this right, are we waiting for someone to clean us up???
```

Comment:
```suggestion
				//TODO(clc): is this right, are we waiting for someone to clean us up???
```

Reply:
@jageall We need to decide on what the correct pattern is here - if the process is terminating I don't know that there's any point in waiting for shutdown here.

Reply:
We should also review other sites that have this kind of shutdown logic in them since this was cargo culted from elsewhere.

Reply:
do we need an issue for this?